File: app/src/main/java/com/example/movies/MovieDetailActivity.java
package com.example.movies;

import static android.content.Intent.ACTION_VIEW;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import io.reactivex.rxjava3.schedulers.Schedulers;

public class MovieDetailActivity extends AppCompatActivity {

    private ImageView imageViewPosterDetail;
    private ImageView imageViewStar;
    private TextView textViewTitle;
    private TextView textViewYear;
    private TextView textViewDescription;
    private RecyclerView recyclerViewTrailers;
    private RecyclerView recyclerViewReviews;
    private static final String MOVIE_KEY = "movie_key";
    private static final String TAG = "MovieDetailActivity";

    private MovieDetailViewModel viewModel;
    private TrailerAdapter trailerAdapter;
    private ReviewAdapter reviewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie_detail);
        initView();
        Movie movie = (Movie) getIntent().getSerializableExtra(MOVIE_KEY);
        viewModel = new ViewModelProvider(this).get(MovieDetailViewModel.class);

        trailerAdapter = new TrailerAdapter();
        recyclerViewTrailers.setAdapter(trailerAdapter);
        viewModel.getTrailers().observe(this,
                trailers -> trailerAdapter.setTrailerList(trailers));
        viewModel.loadTrailer(movie.getId());

        reviewAdapter = new ReviewAdapter();
        recyclerViewReviews.setAdapter(reviewAdapter);
        viewModel.getReviews().observe(this, reviews -> reviewAdapter.setReviewList(reviews));
        viewModel.loadReview(movie.getId());


        Glide.with(this).load(movie.getPoster().getUrl()).into(imageViewPosterDetail);
        textViewTitle.setText(movie.getName());
        textViewYear.setText(String.valueOf(movie.getYear()));
        textViewDescription.setText(movie.getDescription());

        trailerAdapter.setOnTrailerClickListener(trailer -> {
            Intent intent = new Intent(ACTION_VIEW);
            intent.setData(Uri.parse(trailer.getUrl()));
            startActivity(intent);
        });

        Drawable startOn = ContextCompat.getDrawable(this, android.R.drawable.btn_star_big_on);
        Drawable startOff = ContextCompat.getDrawable(this, android.R.drawable.btn_star_big_off);
        viewModel.getFavoriteMovie(movie.getId()).observe(this, favoriteMovie -> {
            if (favoriteMovie == null) {
                imageViewStar.setImageDrawable(startOff);
                imageViewStar.setOnClickListener(l -> viewModel.insertMovie(movie));
            } else {
                imageViewStar.setImageDrawable(startOn);
                imageViewStar.setOnClickListener(l -> viewModel.removeMovie(movie.getId()));
            }
        });
    }

    private void initView() {
        imageViewPosterDetail = findViewById(R.id.imageViewPosterDetail);
        textViewTitle = findViewById(R.id.textViewTitle);
        textViewYear = findViewById(R.id.textViewYear);
        textViewDescription = findViewById(R.id.textViewDescription);
        recyclerViewTrailers = findViewById(R.id.recyclerViewTrailers);
        recyclerViewReviews = findViewById(R.id.recyclerViewReviews);
        imageViewStar = findViewById(R.id.imageViewStar);
    }

    public static Intent newIntent(Context context, Movie movie) {
        Intent intent = new Intent(context, MovieDetailActivity.class);
        intent.putExtra(MOVIE_KEY, movie);
        return intent;
    }
}

File: app/src/main/java/com/example/movies/MovieDetailActivity.kt
package com.example.movies

import android.content.Context
import android.content.Intent
import android.content.Intent.ACTION_VIEW
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import com.bumptech.glide.Glide
import com.example.movies.adapters.ReviewAdapter
import com.example.movies.adapters.TrailerAdapter
import com.example.movies.databinding.ActivityMovieDetailBinding
import com.example.movies.pojo.Movie

class MovieDetailActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMovieDetailBinding
    private lateinit var viewModel: MovieDetailViewModel
    private val trailerAdapter = TrailerAdapter()
    private val reviewAdapter = ReviewAdapter()
    private lateinit var movie: Movie
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_movie_detail)
        binding = ActivityMovieDetailBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProvider(this)[MovieDetailViewModel::class.java]
        movie = intent.getSerializableExtra(MOVIE_KEY) as Movie
        installAdapters()
        observes()
        customizingViews()
    }

    private fun customizingViews() {
        Glide.with(this).load(movie.poster.url).into(binding.imageViewPosterDetail)
        binding.textViewTitle.text = movie.name
        binding.textViewYear.text = movie.year.toString()
        binding.textViewDescription.text = movie.description
    }

    private fun observes() {
        viewModel.getReviews.observe(this) { reviewAdapter.reviewList = it }
        viewModel.getTrailers.observe(this) { trailerAdapter.trailerList = it }
        viewModel.loadReview(movie.id)
        viewModel.loadTrailer(movie.id)

        val startOn = ContextCompat.getDrawable(this, android.R.drawable.btn_star_big_on)
        val startOff = ContextCompat.getDrawable(this, android.R.drawable.btn_star_big_off)
        viewModel.getFavoriteMovie(movie.id).observe(this) {
            if (it == null) {
                binding.imageViewStar.setImageDrawable(startOff)
                binding.imageViewStar.setOnClickListener { viewModel.insertMovie(movie) }
            } else {
                binding.imageViewStar.setImageDrawable(startOn)
                binding.imageViewStar.setOnClickListener { viewModel.removeMovie(movie.id) }
            }
        }
    }

    private fun installAdapters() {
        binding.recyclerViewTrailers.adapter = trailerAdapter
        binding.recyclerViewReviews.adapter = reviewAdapter
        listener()
    }

    private fun listener() {
        trailerAdapter.onTrailerClickListener = TrailerAdapter.OnTrailerClickListener {
            val intent = Intent(ACTION_VIEW)
            intent.data = Uri.parse(it.url)
            startActivity(intent)
        }
    }

    companion object {
        private const val MOVIE_KEY = "movie_key"
        fun newIntent(context: Context, movie: Movie): Intent {
            val intent = Intent(context, MovieDetailActivity::class.java)
            intent.putExtra(MOVIE_KEY, movie)
            return intent
        }
    }
}